### DISCUSSION TITLE
Add option to relax TLS security for e-mail settings

### DISCUSSION THREAD

**Discussion by michnovka** [btcpayserver/btcpayserver#3895] [Ideas / Feature Requests]
We are using mail server behind VPN tunnel and thus use only self-signed cert. When trying to configure e-mail under latest btcpay server, I get the following error:

![image](https://user-images.githubusercontent.com/16553087/175318851-fbf47fc2-1b33-446e-8f41-c8ad4b3a6f7f.png)

When I tried to use domain name that would match the cert, I still got

![image](https://user-images.githubusercontent.com/16553087/175318947-36c31c87-30fe-4506-8299-ccff12162dff.png)

I would like to add an option to disable these security checks. Users should be able to choose whether it is relevant or ok to ignore. Like I said, our mail server is firewalled and can be accessed only over VPN, so in our use-case this requirement does not improve security at all, the opposite. It actually forces me to open the mail server to the internet and get a commercial cert.

**Comment by NicolasDorier** [comment]
why do you use a certificate then? using a self signed certificate and ignoring the check is completely useless, as it doesn't protect against any attack vector.

**Reply by michnovka** [reply]
This is because some other services we use require the use of certificate, even tho its self-signed. All those however either ignore the self signed cert or allow you to accept an exception.

And I disagree with your conclusion that self signed certs = useless. In most cases the exception for "invalid", "expired" or "self signed" cert is only made for a specific certificate (whose fingerprint is displayed upon accepting the exception) and thus you can validate the cert is indeed yours (kind of cert pinning). If the cert were to change, the exception would be void.

Still I would be happy with just an option to "relax TLS cert checks", and not a concrete cert exception.

**Reply by michnovka** [reply]
FYI I tested to use port 25 (no cert) and it still tries to make TLS connection

![image](https://user-images.githubusercontent.com/16553087/175539077-367c14d5-fef9-443d-aaff-355ca8185722.png)

**Reply by NicolasDorier** [reply]
I agree with cert pinning, should we allow to do that then?

**Reply by NicolasDorier** [reply]
@michnovka for SSL being used on port 25, this is a bug from the lib we are using https://github.com/jstedfast/MailKit/pull/1398

I implemented a workaround for us https://github.com/btcpayserver/btcpayserver/commit/9423bc4ea71747809b029dadfdf3545d5a6d0b68

Can you try type port `0` instead of `25`? Silently the library we use should then use port 25 without SSL.

**Reply by michnovka** [reply]
@NicolasDorier thanks for the tip, but port 0 still tries SSL it seems (at least for the test e-mail):

![image](https://user-images.githubusercontent.com/16553087/175906750-1580eaac-5456-474c-9123-8db29d94aaf9.png)

**Reply by michnovka** [reply]
@NicolasDorier ah, I didnt realize this is a new commit. Since I am using docker set-up I need to wait for a new release. This is a production server, I cannot afford to experiment with it a lot

**Reply by jstedfast** [reply]
Hi,

The SecureSocketOption.StartTlsWhenAvailable option is internally what MailKit uses when the app uses SecureSocketOptions.Auto for port 25 or 0.

Essentially what this means is: if the server claims to support SSL/TLS, then make use of it because it's more secure for the user.

If you don't want that, use SecureSocketOption.None instead of Auto.

**Reply by michnovka** [reply]
Maybe instead of this hack a checkbox to "disable TLS" would be a better option @NicolasDorier ?

**Reply by NicolasDorier** [reply]
thanks for the answer @jstedfast . I think your behavior make sense.
@michnovka note that TLS checks are disabled if the server is local, where local is defined by:
```csharp
        public static bool IsLocalNetwork(string server)
        {
            ArgumentNullException.ThrowIfNull(server);
            if (Uri.CheckHostName(server) == UriHostNameType.Dns)
            {
                return server.EndsWith(".internal", StringComparison.OrdinalIgnoreCase) ||
                   server.EndsWith(".local", StringComparison.OrdinalIgnoreCase) ||
                   server.EndsWith(".lan", StringComparison.OrdinalIgnoreCase) ||
                   server.IndexOf('.', StringComparison.OrdinalIgnoreCase) == -1;
            }
            if (IPAddress.TryParse(server, out var ip))
            {
                return ip.IsLocal() || ip.IsRFC1918();
            }
            return false;
        }
```

**Reply by NicolasDorier** [reply]
Made this PR https://github.com/btcpayserver/btcpayserver/pull/3908

**Comment by NicolasDorier** [comment]
Fixed by https://github.com/btcpayserver/btcpayserver/pull/3908
